tests/runtime-tests/src/lib.rs

```diff
@@ -0,0 +1,400 @@
+            r#"malformed response does not end with the expected CRLF"#
+        );
+        if code == 500 {
+            let header_end = s
+                .find("\r\n\r\n")
+                .context("malformed response does not contain CRLF header separator")?;
+            if header_end + 4 == s.len() {
+                return Ok(Response::ErrNoBody);
+            }
+            let body =
+                String::from_utf8(s.as_bytes()[header_end + 4..s.len() - 4].to_vec()).unwrap();
+            return Ok(Response::Err(body));
+        }
+        if code == 200 {
+            return Ok(Response::Ok);
+        }
+        anyhow::bail!("Could not parse HTTP raw response: {s}")
+    }
+}
+
+fn make_http_request(spin: &mut Spin) -> Result<String, anyhow::Error> {
+    if let Some(status) = spin.try_wait()? {
+        anyhow::bail!("Spin exited early with status code {:?}", status.code());
+    }
+    log::debug!("Connecting to HTTP server on port {}...", spin.port());
+    let mut connection = std::net::TcpStream::connect(format!("127.0.0.1:{}", spin.port()))
+        .context("could not connect to Spin web server")?;
+    connection.write_all(b"GET / HTTP/1.1\r\nHost: 127.0.0.1:3000\r\n\r\n")?;
+    log::debug!("Awaiting response from server");
+    let mut start = 0;
+    let mut output = vec![0; 1024];
+    for _ in 0..20 {
+        let n = connection.read(&mut output[start..])?;
+        start += n;
+        let header_end = output.windows(4).position(|c| c == b"\r\n\r\n");
+        if let Some(header_end) = header_end {
+            let until_headers = String::from_utf8(output[..header_end].to_vec())
+                .context("spin HTTP response headers contained non-utf8 bytes")?;
+            if let Some(s) = until_headers.find("content-length: ") {
+                if std::str::from_utf8(&until_headers.as_bytes()[s + 16..])
+                    .unwrap()
+                    .starts_with('0')
+                {
+                    let response_with_no_body =
+                        String::from_utf8(output[..header_end + 4].to_vec()).unwrap();
+                    return Ok(response_with_no_body);
+                }
+            }
+            if output[header_end + 4..]
+                .windows(4)
+                .any(|c| c == b"\r\n\r\n")
+            {
+                let response = String::from_utf8(output[..start].to_vec())
+                    .context("spin HTTP response contained non-utf8 bytes")?;
+                return Ok(response);
+            }
+        }
+        std::thread::sleep(std::time::Duration::from_millis(250));
+        if let Some(status) = spin.try_wait()? {
+            anyhow::bail!("Spin exited early with status code {:?}", status.code());
+        }
+    }
+    anyhow::bail!("did not return http response after 5 seconds")
+}
+
+struct Spin {
+    process: std::process::Child,
+    #[allow(dead_code)]
+    stdout: OutputStream,
+    stderr: OutputStream,
+    port: u16,
+}
+
+impl Spin {
```

Comment:
could we change ths to a trait and have a Spin implementation (and then a cloud implementation). that way we can try to reuse same tests for cloud tests.